cps/elf.c

```diff
@@ -0,0 +1,261 @@
+
+	/* Length of file contents. */
+	unsigned long len;
+
+	/* Whether the endianness needs to be converted. */
+	int  conv;
+};
+
+/*
+ * Get CPU endianness.
+ *
+ * 0 = unknown
+ * 1 = ELFDATA2LSB = little
+ * 2 = ELFDATA2MSB = big
+ */
+static int __attribute__((pure))
+native_endianness(void)
+{
+	/*
+	 * Encoding the endianness enums in a string
+	 * and then reading that string as a 32-bit int,
+	 * returns the correct endianness automagically.
+	 */
+	return (char)*((const uint32_t *)("\1\0\0\2"));
+}
+
+static inline void
+__swap_bytes(const void *src, void *dest, unsigned int size)
+{
+	unsigned int i;
+	for (i = 0; i < size; i++) {
+		((unsigned char *)dest)[i] =
+			((const unsigned char *)src)[size - i - 1];
+	}
+}
+
+/* Change endianness of x if @conv is true. */
+#define END(x, conv)							\
+({									\
+	typeof(x) __x;							\
+	if (conv) __swap_bytes(&(x), &(__x), sizeof(__x));		\
+	else __x = (x);							\
+	__x;								\
+})
+
+/* Get the section specified in @secname from the ELF file. */
+static void *
+PERBIT(get_section)(struct elf_file *mod, const char *secname,
+	unsigned long *secsize)
+{
+	void *data = mod->data;
+	unsigned long len = mod->len;
+	int conv = mod->conv;
+	unsigned int i;
+	const char *secstrings;
+
+	ElfPERBIT(Ehdr) *hdr = data;
+	ElfPERBIT(Shdr) *sechdrs;
+	ElfPERBIT(Off) e_shoff;
+	ElfPERBIT(Half) e_shnum, e_shstrndx;
+
+	*secsize = 0;
+
+	/* Check whether there's enough room for ELF header. */
+	if (len < sizeof(*hdr))
+		return NULL;
+
+	/* Read in necessary variables from ELF header and fix endianness. */
+	e_shoff = END(hdr->e_shoff, conv);
+	e_shnum = END(hdr->e_shnum, conv);
+	e_shstrndx = END(hdr->e_shstrndx, conv);
+
+	/* Check whether there's enough room for the section headers. */
+	if (len < e_shoff + e_shnum * sizeof(sechdrs[0]))
+		return NULL;
+
+	sechdrs = (ElfPERBIT(Shdr) *)((uint8_t *)hdr + e_shoff);
+
+	/* Check whether there's enough room for section with section names. */
+	if (len < END(sechdrs[e_shstrndx].sh_offset, conv))
+		return NULL;
+
+	/* Fix endianness in section headers if needed. */
+	for (i = 0; i < e_shnum; i++) {
+		sechdrs[i].sh_name = END(sechdrs[i].sh_name, conv);
+		sechdrs[i].sh_offset = END(sechdrs[i].sh_offset, conv);
+		sechdrs[i].sh_size = END(sechdrs[i].sh_size, conv);
+	}
+
+	/* Find symbol table. */
+	secstrings = (const char *)
+		((uint8_t *)hdr + sechdrs[e_shstrndx].sh_offset);
+	for (i = 1; i < e_shnum; i++) {
+		const char *found_secname = secstrings + sechdrs[i].sh_name;
+		if (strcmp(secname, found_secname) == 0) {
+			ElfPERBIT(Off) secoffset = sechdrs[i].sh_offset;
+
+			/* Not enough room in file for this section. */
+			if (len < secoffset + sechdrs[i].sh_size)
+				return NULL;
+
+			*secsize = sechdrs[i].sh_size;
+			return (uint8_t *)hdr + secoffset;
+		}
+	}
+
+	return NULL;
+}
+
+/* Find the next string in an ELF section. */
+static const char *
+next_string(const char *string, unsigned long *secsize)
+{
+	/* Skip non-zero chars. */
```

Comment:
Add the following code to make sure that `next_string()` is safe when when it's called with `*secsize == 0`.
```
if (*secsize == 0)
	return NULL;
```